lib/vquic/ngtcp2.c

```diff
@@ -717,20 +717,62 @@ static int cb_h3_stream_close(nghttp3_conn *conn, int64_t stream_id,
   return 0;
 }
 
+/*
+ * h3_buffer_grow() makes room for new data at the end of the receive buffer.
+ * We make sure that the full data fits in the allocated header buffer, or
+ * else we enlarge it. We need to do this because we can't limit HTTP/3
+ * headers precisely via QUIC flow control.
+ */
+static CURLcode h3_buffer_grow(struct Curl_easy *data,
+                               struct HTTP *stream,
+                               size_t length)
+{
+  struct SingleRequest *k = &data->req;
+  /* length can be arbitrarily large, so take care not to overflow. */
+  size_t maxleft = CURL_MAX_READ_SIZE - stream->memlen;
+  if(length > maxleft) {
+    /* The reason to have a max limit for this is to avoid the risk of a bad
+       server feeding libcurl with a highly compressed list of headers that
+       will cause our buffer to grow too large */
+    failf(data, "Rejected %zu bytes list of headers (max is %d)!",
+          stream->memlen + length, CURL_MAX_READ_SIZE);
+    return CURLE_OUT_OF_MEMORY;
+  }
+  size_t newsize = stream->memlen + length;
+  if(newsize >= data->set.buffer_size) {
+    /* We enlarge the receive buffer as it is too small */
+    char *newbuff;
+    newsize = CURLMAX(newsize * 3 / 2, data->set.buffer_size*2);
+    newsize = CURLMIN(newsize, CURL_MAX_READ_SIZE);
+    newbuff = realloc(k->buf, newsize + 1);
+    if(!newbuff) {
+      failf(data, "Failed to alloc memory for resized receive buffer!");
+      return CURLE_OUT_OF_MEMORY;
+    }
+    /* Update all the pointers to this buffer now that it's reallocated */
+    data->state.buffer = k->buf = newbuff;
```

Comment:
I'm not happy about the fact that this code is assuming that the buffer passed to `ngh3_stream_recv` is the request receive buffer. It seems to always be the case, but it breaks encapsulation. Should we do something about this?